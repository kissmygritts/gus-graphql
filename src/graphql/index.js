import { fileLoader, mergeTypes, mergeResolvers } from 'merge-graphql-schemas'
import path from 'path'

const typesArray = fileLoader(path.join(__dirname, './**/*.schema.graphql'), { recursive: true })
const resolversArray = fileLoader(path.join(__dirname, './**/*.resolver.js'), { recursive: true })

export const typeDefs = mergeTypes(typesArray)
export const resolvers = mergeResolvers(resolversArray)
