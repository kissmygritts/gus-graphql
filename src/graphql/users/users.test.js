import 'dotenv/config'
import { graphql } from 'graphql'
import { makeExecutableSchema } from 'graphql-tools'
import { typeDefs, resolvers } from '..'
// import { db } from '../../db'
import { cleanDatabase } from '../../../test/helper.js'

const schema = makeExecutableSchema({
  typeDefs,
  resolvers
})

const signupMutation = /* GraphQL */ `
  mutation($input: SignUpInput) {
    signup(input: $input) {
      token
    }
  }
`

const signinMutation = /* GraphQL */ `
  mutation($input: SignInInput) {
    signin(input: $input) {
      token
    }
  }
`

const queryRunner = async (query, variables) => {
  return graphql(schema, query, null, {}, variables)
}

const username = 'mitch'
const email = 'mitch@email.com'
const password = 'password'

describe('Signup new users', () => {
  beforeAll(() => cleanDatabase())

  // afterAll(() => {
  //   cleanDatabase().then(() => db.$pool.end())
  // })

  it('return a token', async () => {
    const response = await queryRunner(signupMutation, {
      input: { username, email, password }
    })
    expect(typeof response.data.signup.token).toBe('string')
  })

  it("return error when password isn't a string", async () => {
    const response = await queryRunner(signupMutation, {
      input: { username, email, password: 1234 }
    })
    expect(response).toMatchSnapshot()
  })

  it("return error when email isn't a string", async () => {
    const response = await queryRunner(signupMutation, {
      input: { username, email: 1234, password }
    })
    expect(response).toMatchSnapshot()
  })

  it("return error when username isn't a string", async () => {
    const response = await queryRunner(signupMutation, {
      input: { username: 1234, email, password }
    })
    expect(response).toMatchSnapshot()
  })

  it('return duplicate entry error when same email is used', async () => {
    const response = await queryRunner(signupMutation, {
      input: { username: 'tim', email, password }
    })
    expect(response).toMatchSnapshot()
  })

  it('return duplicate entry error when same username is used', async () => {
    const response = await queryRunner(signupMutation, {
      input: { username, email: 'tim@email', password }
    })
    expect(response).toMatchSnapshot()
  })

  it('return null constraint error when a null value is used', async () => {
    // this validation isn't passing the graphql validation step because
    // it is an invalid query: https://graphql.org/learn/validation/
    const response = await queryRunner(signupMutation, {
      input: { username: null, email, password }
    })
    expect(response).toMatchSnapshot()
  })
})

describe('Signin a user', () => {
  beforeAll(() => {
    return queryRunner(signupMutation, {
      input: { username, email, password }
    })
  })

  it('return a token after successful signin', async () => {
    const response = await queryRunner(signinMutation, {
      input: { email, password }
    })
    expect(typeof response.data.signin.token).toBe('string')
  })

  it('return an error if user is not found', async () => {
    const response = await queryRunner(signinMutation, {
      input: { email: 'mike@email.com', password }
    })
    expect(response).toMatchSnapshot()
  })

  it('return an error if password is incorrect', async () => {
    const response = await queryRunner(signinMutation, {
      input: { email, password: 'pass' }
    })
    expect(response).toMatchSnapshot()
  })
})
