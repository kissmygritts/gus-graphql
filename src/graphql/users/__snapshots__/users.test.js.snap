// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signin a user return an error if password is incorrect 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Invalid password],
  ],
}
`;

exports[`Signin a user return an error if user is not found 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: No user found with this emaill address.],
  ],
}
`;

exports[`Signup new users return duplicate entry error when same email is used 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: duplicate key value violates unique constraint "users_email_key"],
  ],
}
`;

exports[`Signup new users return duplicate entry error when same username is used 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: duplicate key value violates unique constraint "users_username_key"],
  ],
}
`;

exports[`Signup new users return error when email isn't a string 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$input" got invalid value { username: "mitch", email: 1234, password: "password" }; Expected type String at value.email; String cannot represent a non string value: 1234],
  ],
}
`;

exports[`Signup new users return error when password isn't a string 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$input" got invalid value { username: "mitch", email: "mitch@email.com", password: 1234 }; Expected type String at value.password; String cannot represent a non string value: 1234],
  ],
}
`;

exports[`Signup new users return error when username isn't a string 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$input" got invalid value { username: 1234, email: "mitch@email.com", password: "password" }; Expected type String at value.username; String cannot represent a non string value: 1234],
  ],
}
`;

exports[`Signup new users return null constraint error when a null value is used 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Variable "$input" got invalid value { username: null, email: "mitch@email.com", password: "password" }; Expected non-nullable type String! not to be null at value.username.],
  ],
}
`;
